src/Model.integration.test.ts

```diff
@@ -0,0 +1,244 @@
+    name: string;
+    data: any;
+    mutationId: string;
+  };
+}
+
+describe('Stream integration', () => {
+  beforeEach<TestContext>(async (context) => {
+    const channelName = 'test-channel';
+    const data = await createAblyApp({
+      keys: [{}],
+      namespaces: [{ id: channelName, persisted: true }],
+      channels: [
+        {
+          name: channelName,
+          presence: [
+            { clientId: 'John', data: 'john@test.com' },
+            { clientId: 'Dave', data: 'dave@test.com' },
+          ],
+        },
+      ],
+    });
+    const ably = new Realtime({
+      key: data.keys[0].keyStr,
+      environment: 'sandbox',
+    });
+    const logger = pino({ level: 'debug' });
+    const model = new Model(
+      channelName,
+      {
+        sync: async (data) => data,
+        merge: (state, event) => (state ? { ...state, ...event.data } : event.data),
+      },
+      {
+        ably,
+        channelName,
+        logger,
+        syncOptions: defaultSyncOptions,
+        optimisticEventOptions: defaultOptimisticEventOptions,
+        eventBufferOptions: defaultEventBufferOptions,
+      },
+    );
+
+    context.model = model;
+    context.channelName = channelName;
+    context.ably = ably;
+
+    context.eventData = {
+      name: 'update',
+      data: {
+        noOfMessages: 34,
+      },
+      mutationId: 'some-id-1',
+    };
+  });
+
+  afterEach<TestContext>(async ({ model }) => {
+    await model.dispose();
+    vi.restoreAllMocks();
+  });
+
+  it<TestContext>('changes state on sync, pause, resume, dispose', async ({ model }) => {
+    expect(model.state).toEqual('initialized');
+
+    const sync = model.sync({
+      data: {
+        foo: 'foobar',
+        bar: {
+          baz: 1,
+        },
+      },
+      sequenceId: '1',
+    });
+    expect(model.state).toEqual('syncing');
+
+    await sync;
+    expect(model.state).toEqual('ready');
+
+    await model.pause();
+    expect(model.state).toEqual('paused');
+
+    await model.resume();
+    expect(model.state).toEqual('ready');
+
+    await model.dispose();
+    expect(model.state).toEqual('disposed');
+  });
+
+  it<TestContext>('successfully sets the data from the event in optimistic()', async ({
+    ably,
+    model,
+    channelName,
+    eventData,
+  }) => {
+    await model.sync(eventData.data);
+
+    let subscription = new Subject<void>();
+    const subscriptionCalls = getEventPromises(subscription, 3);
+    const subscriptionSpy = vi.fn(() => subscription.next());
+
+    model.subscribe(subscriptionSpy);
+
+    const [confirmation] = await model.optimistic(eventData);
+
+    expect(model.data.optimistic).toEqual(eventData.data);
+
+    const channel = ably.channels.get(channelName);
+    await channel.publish({
+      data: eventData.data,
+      name: 'update',
+      extras: {
+        headers: {
+          'x-ably-models-event-uuid': eventData.mutationId,
+        },
+      },
+    });
+
+    await subscriptionCalls[2];
```

Comment:
We should also await `subscriptionCalls[0]` on L119 and assert initial state, then `subscriptionCalls[1]` on L132 and assert optimistic state, and finally `subscriptionCalls[2]` and assert confirmed state

Reply:
when you say "initial state" do you mean `model.state`? Because `model.data` only has `optimistic` and `confirmed` doesn't it?

Reply:
we're also currently asserting only on the value of `model.data.confirmed` rather than the value the the subscription callback is actually invoked with, that would be worth asserting on too